UI Divider

alex40724 opened the pull request:
Separates thematic content. E.g. being used in future Dropdowns.
Related discussions: https://github.com/ILIAS-eLearning/ILIAS/pull/544 and https://github.com/ILIAS-eLearning/ILIAS/pull/550

Commit:
ks divider

Amstutz commented on src/UI/Component/Divider/Factory.php:
```diff
@@ -0,0 +1,32 @@
+	 *   ordering:
+	 *       1: >
+	 *          Dividers MUST always have a preceding and a succeeding element
+	 *          in a sequence of elments, which MUST NOT be another Divider.
```
Good! Maybe add the full name "Standard Dividers MUST always ... ", just to avoid any confusion with the general family of dividers.

Amstutz commented on src/UI/Factory.php:
```diff
@@ -336,4 +336,22 @@ public function panel();
+	 * rules:
+	 *   usage:
+	 *     1: >
+	 *       Dividers MUST be used in in container components that render a sequence of items.
```
Is it a) MUST be used or b) MUST only be used? If a) I disagree. There can be dropdowns without dividers. If b) add only the clarify.

Amstutz commented on src/UI/templates/default/Divider/divider.less:
```diff
@@ -0,0 +1,13 @@
+/* not final */
+h6.il-divider {
+	padding: 3px 10px;
+	background-color: #f0f0f0;
```
No color key in less files other than the variable.less, see: https://github.com/ILIAS-eLearning/ILIAS/blob/trunk/templates/Readme.md#colors. Either use an existing color value directly or add an new one for dividers, e.g. divider-bg.

Amstutz commented on src/UI/templates/default/Divider/divider.less:
```diff
@@ -0,0 +1,13 @@
+/* not final */
+h6.il-divider {
+	padding: 3px 10px;
```
Consider using: padding-small-vertical and padding-small-horizontal here.

Amstutz commented on src/UI/templates/default/Divider/divider.less:
```diff
@@ -0,0 +1,13 @@
+	padding: 3px 10px;
+	background-color: #f0f0f0;
+	display: block;
+	font-size: 12px;
```
Why not variable: font-size-base

Amstutz commented on src/UI/templates/default/Divider/divider.less:
```diff
@@ -0,0 +1,13 @@
+	background-color: #f0f0f0;
+	display: block;
+	font-size: 12px;
+	line-height: 1.42857143;
```
Use line-height-base

Amstutz commented on src/UI/templates/default/Divider/divider.less:
```diff
@@ -0,0 +1,13 @@
+	display: block;
+	font-size: 12px;
+	line-height: 1.42857143;
+	color: #777777;
```
No color key in less files other than the variable.less, see: https://github.com/ILIAS-eLearning/ILIAS/blob/trunk/templates/Readme.md#colors. Either use an existing color value directly or add an new one for dividers, e.g. divider-text-color or divider-color.

Amstutz reviewed (commented):
Thank you very much looks good and is JF ready. 

Before merging this:
- Add the according tests.
- Adapt the less file by using the existing variable or create new ones (for coloring) if necessary.

Commit:
rule clarifications

alex40724 commented on src/UI/Factory.php:
```diff
@@ -336,4 +336,23 @@ public function panel();
+	 *   purpose: >
+	 *     A divider marks a thematic change in a sequence of other components.
+	 *   composition: >
+	 *     A divider consists of a horizontal ruler which may be accompanied by a label.
```
JF: ...and may comprise a label...

chfsx commented on src/UI/Factory.php:
```diff
@@ -336,4 +336,23 @@ public function panel();
+	 * rules:
+	 *   usage:
+	 *     1: >
+	 *       Dividers MUST only be used in in container components that render
```
typo with double `in in`. 

I think the rule needs some improvement since i think it's currently to weak and I as a developer do not see that I should not use it vor horizontal listings.

chfsx commented on src/UI/Factory.php:
```diff
@@ -336,4 +336,23 @@ public function panel();
+	 *
+	 * @return \ILIAS\UI\Component\Divider\Factory
+	 **/
+	public function divider();
```
This seems to be too generic for me, since it's only a horizontal divider, but there could be vertical as well.

alex40724 replied:
I renamed the element to Divider > Horizontal and changed the description.

chfsx commented on src/UI/Implementation/Component/Divider/Standard.php:
```diff
@@ -0,0 +1,40 @@
+	/**
+	 * @inheritdoc
+	 */
+	public function withLabel($label) {
```
This makes only sende in horizontal dividers.

matthiaskunkel commented:
Jour Fixe, June 19, 2017: Pull request accepted. Please merge to trunk.

Commit:
renamed standard to horizontal, clarified description, added tests, fixed less

alex40724 commented:
I
- renamed standard to horizontal and clarified description due to @chfsx comments
- added tests and fixed less due to @Amstutz  comments

So it's $f->divider()->horizontal() instead of $f->divider()->standard() now. I agree with @chfsx that a vertical divider would be a separate subcomponent (not implemented yet).

@matthiaskunkel , @Amstutz Can this be moved to the trunk in this way?

Does anybody else object these changes? It would be nice to get some feedback, since I would not like to schedule another JF slot for this. If someone thinks this is necessary, please add a comment.

Commit:
removed comment

matthiaskunkel reviewed (approved):
Fine for me.

Amstutz commented on src/UI/Component/Divider/Factory.php:
```diff
@@ -0,0 +1,39 @@
+	 * ---
+	 * description:
+	 *   purpose: >
+	 *       Horizontal Divider is used to mark a thematic change in sequence of
```
**A** Horizontal ... or Horizontal Dividers are ...

Amstutz commented on src/UI/templates/default/Divider/divider.less:
```diff
@@ -0,0 +1,12 @@
+	display: block;
+	font-size: @font-size-small;
+	line-height: @line-height-small;
+	color: lighten(@text-color, 10%);
```
Do not use lighten in any other files than variables.less. See: https://github.com/ILIAS-eLearning/ILIAS/tree/trunk/templates#colors

Note that this is otherwise extremely hard to track for designers. We had just recent complaints about this. 

Either introduce an il-divider-color or an il-text-color-light, with an according description of where you intend to use this.

Amstutz commented on src/UI/templates/default/Divider/divider.less:
```diff
@@ -0,0 +1,12 @@
+h6.il-divider {
+	padding: @padding-small-vertical @padding-small-horizontal;
+	background-color: darken(@il-secondary-container-bg, 5%);
```
Do not use lighten in any other files than variables.less. See: https://github.com/ILIAS-eLearning/ILIAS/tree/trunk/templates#colors

Introduce an il-divider-bg color for this.

Amstutz reviewed (changes requested):
LGTM, However there is one little less change. Then this is ready to merge. I do not think we need an other JF discussion for this.

Commit:
fixed description and less variables
